frame/balances/src/migration.rs

```diff
@@ -15,50 +15,57 @@
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
 use super::*;
-use frame_support::{pallet_prelude::*, traits::OnRuntimeUpgrade, weights::Weight};
+use frame_support::{
+	pallet_prelude::*,
+	traits::{OnRuntimeUpgrade, PalletInfoAccess},
+	weights::Weight,
+};
 
-// NOTE: This must be used alongside the account whose balance is expected to be inactive.
-// Generally this will be used for the XCM teleport checking account.
-pub struct MigrateToTrackInactive<T, A>(PhantomData<(T, A)>);
-impl<T: Config, A: Get<T::AccountId>> OnRuntimeUpgrade for MigrateToTrackInactive<T, A> {
-	fn on_runtime_upgrade() -> Weight {
-		let current_version = Pallet::<T>::current_storage_version();
-		let onchain_version = Pallet::<T>::on_chain_storage_version();
+fn migrate_v0_to_v1<T: Config<I>, I: 'static>(accounts: &[T::AccountId]) -> Weight {
+	let onchain_version = Pallet::<T, I>::on_chain_storage_version();
+
+	if onchain_version == 0 {
```

Comment:
Does the migration or the new pallet get applied first? Because in the pallet you set the storage version to `1`, so if that applies first then IIUC this migration won't actually run.
```rust
/// The current storage version.
const STORAGE_VERSION: frame_support::traits::StorageVersion =
	frame_support::traits::StorageVersion::new(1);
```

Reply:
On chain version will return 0, at this point this storage item even does not exist.

here the trait
https://github.com/paritytech/substrate/blob/a1d014d3489726e4dbb6bd46e164bdf83242e5ec/frame/support/src/traits/metadata.rs#L242

here the impl 
https://github.com/paritytech/substrate/blob/a1d014d3489726e4dbb6bd46e164bdf83242e5ec/frame/support/src/dispatch.rs#L2518

and below here in the migration you can see where we updating on chain storage version
https://github.com/paritytech/substrate/pull/12840/files#diff-d3a8ee05060d4cc1bb98df686418ec5aeba54698bd793a46c9095dde576eaa77R41

Reply:
The storage version is only set on genesis to the "current version" (aka the version you set in the code). After genesis, you need to bump the version "manually".

Reply:
@joepetrowski that constant, is not persisted anywhere by frame, its only returned by on_chain_storage_version
here 
https://github.com/paritytech/substrate/blob/a1d014d3489726e4dbb6bd46e164bdf83242e5ec/frame/support/src/dispatch.rs#L2522